Admin Generator (Future): Forward required GraphQL args for forms

Ben-Ho opened the pull request:
Allows setting required GraphQL args from outside of the generated form. For instance, creating depending relation-data.

johnnyomair commented:
How do you determine which fields should be forwarded? By comparing the fields in the config with the GraphQL input and use what's missing?

Ben-Ho commented on demo/admin/src/products/future/generated/CreateCapProductForm.tsx:
```diff
@@ -15,35 +15,39 @@ import {
+    type: GQLProductType;
+}
+
+export function CreateCapProductForm({ type }: FormProps): React.ReactElement {
```
we need more demo-code. I didn't considered if the form is also used to edit the entity. in this case this additional props are not required, because they only apply for create. so this props should require id or required props... @nsams @johnnyomair should I address this in another pull-request?

nsams replied:
are you thinking of `{ id: string } | { type: GQLProductType }`? Does this even work?

Ben-Ho replied:
it seems this is possible. I found an article describing the solution is defining never for the other props. But this does only validate the type does not match but does not help with autocomplete.

```
type FormProps =
    | {
        id: string;
        type?: never;
      }
    | {
        id?: never;
        type: GQLProductType;
      };
```

my other solution would be not working with root-props but with nested

```
type FormProps = {
    data: {
        id: string;
    } | {
        type: GQLProductType;
    }
}
```

Ben-Ho commented:
> maybe I should compare field-config only with args in input-arg, because it shouldn't be too complicated and field-config will probably never contain root-args

I changed it to filter only input-arg fields, because config-fields are always part of the input-arg.

and I removed the "skip id" because id is only used for the update-mutation which does currently not need a forward required-gql-args feature.

Ben-Ho commented:
@nsams @johnnyomair I think this pull-request is ready. changing the way props are defined should be changed in another pull-request if we want to.

johnnyomair commented on packages/admin/cms-admin/src/generator/future/generateForm.ts:
```diff
@@ -89,10 +128,28 @@ export function generateForm(
-            mutation Create${gqlType}($input: ${gqlType}Input!) {
-                ${createMutation.name}(input: $input) {
+            mutation Create${gqlType}(${
+            gqlArgs.filter((gqlArg) => !gqlArg.isInputArgSubfield && gqlArg.queryOrMutationName === `create${gqlType}`).length
+                ? `${gqlArgs
+                      .filter((gqlArg) => !gqlArg.isInputArgSubfield && gqlArg.queryOrMutationName === `create${gqlType}`)
+                      .map((gqlArg) => {
+                          return `$${gqlArg.name}: ${gqlArg.type}!`;
+                      })
+                      .join(", ")}, `
+                : ``
```
You don't need this length checks here, since joining an empty string array results in an empty string anyway

Ben-Ho replied:
it's about this unremarkable comma after join. .join(", ")}**,** ` this would be too much... honestly I'm no fan of this string-concatenation stuff. I would prefer a function generating this mutation-params string hiding away 20 lines of code... but this will hopefully part of a later refactoring :-)

johnnyomair commented on packages/admin/cms-admin/src/generator/future/generateForm/getForwardedGqlArgs.ts:
```diff
@@ -0,0 +1,98 @@
+            props.push({ name: arg.name, optional: false, type: `GQL${arg.type}` }); // generated types contain GQL prefix
+            imports.push({ name: `GQL${arg.type}`, importPath: "@src/graphql.generated" });
+        }
+        gqlArgs.push({ name: arg.name, type: arg.type, queryOrMutationName: arg.gqlField.name, isInputArgSubfield: arg.isInputArgSubfield });
```
Why do I need `queryOrMutationName`? Shouldn't the code be applied to a single GraphQL operation anyway?

Ben-Ho replied:
because initially I checked the create and update mutations so I needed to know the origin of this argument to apply it to the correct mutations. additionally generate grid uses practically the same gqlArgs idea with the same type-property-naming therefore the naming **query**OrMutationName...

In a later refactoring I could rename this field to operationName, because I learned query and mutations are both operations...

I didn't came up with an idea where an update-mutation could use a required gql arg so I decided to don't check the update-mutation for now.

Do you want me to remove this field?

johnnyomair replied:
What do you think about passing `gqlOperation` as argument for `getForwardedGqlArgs` instead? Then the function could be called twice for create and update mutations (if ever needed), and we could remove the filter of the args based on the operation name. IMO this would be easier to understand.

Ben-Ho replied:
good idea, required special duplicate-handling either way needs to be done where the code is rendered like in `generateFormPropsCode(...)`.

Ben-Ho replied:
> What do you think about passing gqlOperation as argument

`gqlField` is already the gqlOperation, but the type of an gqlOperation is `IntrospectionField` so I named it `gqlField`. But I think `gqlOperation` as name fits better.

johnnyomair commented:
@Ben-Ho please fix conflicts (sorry)

Ben-Ho commented:
> @Ben-Ho please fix conflicts (sorry)

no problem, simple import conflicts. solved 👍 (I hope I didn't introduce any linting problems :-D)
